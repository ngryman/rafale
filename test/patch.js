import test from 'ava'
import { nodeToJSON } from './helpers/to-json'
import { html, mount, diff, patch } from '../'

function macro(t, newVTree, oldVTree, html) {
  mount(oldVTree, t.context)
  patch(diff(newVTree, oldVTree))
  t.snapshot(nodeToJSON(t.context))
  t.true(newVTree.domNode.isSameNode(t.context.firstElementChild))
}

const renderPrimitive = (text) => text
const renderBold = (text) => html`<b>${text}</b>`
const renderItalic = (text) => html`<i>${text}</i>`
const wrap = (content) => html`<p>${content}</p>`

test.beforeEach(t => {
  const node = document.createElement('div')
  document.body.appendChild(node)
  t.context = node
})

test.afterEach(t => {
  t.context.remove()
})

test('morph a root component', macro,
  renderItalic('foo'),
  renderBold('foo')
)

test('morph a component', macro,
  wrap(renderItalic('foo')),
  wrap(renderBold('foo')),
)

test('morph a component into a primitive', macro,
  wrap(renderPrimitive('foo')),
  wrap(renderBold('foo'))
)

test('morph a component into a sequence', macro,
  wrap(['foo', 'bar'].map(renderBold)),
  wrap(renderBold('foo'))
)

test('morph a sequence', macro,
  wrap(['foo', 'bar'].map(renderItalic)),
  wrap(['foo', 'bar'].map(renderBold))
)

test('morph a sequence into a component', macro,
  wrap(renderBold('foo')),
  wrap(['foo', 'bar'].map(renderBold))
)

test('morph a sequence into a primitive', macro,
  wrap(renderPrimitive('foo')),
  wrap(['foo', 'bar'].map(renderBold))
)

test('add items to a sequence', macro,
  wrap(['foo', 'bar', 'baz', 'qux'].map(renderBold)),
  wrap(['foo', 'bar'].map(renderBold))
)

test('remove items from a sequence', macro,
  wrap(['foo', 'baz'].map(renderBold)),
  wrap(['foo', 'bar', 'baz', 'qux'].map(renderBold))
)

test('clear a sequence', macro,
  wrap([].map(renderBold)),
  wrap(['foo', 'bar'].map(renderBold))
)

test('update items of a sequence', macro,
  wrap(['baz', 'qux'].map(renderBold)),
  wrap(['foo', 'bar'].map(renderBold))
)

test('reorder items of a sequence', macro,
  wrap(['qux', 'baz', 'bar', 'foo'].map(renderBold)),
  wrap(['foo', 'bar', 'baz', 'qux'].map(renderBold))
)
