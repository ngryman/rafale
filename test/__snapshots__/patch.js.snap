// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add items to a sequence 1`] = `
Object {
  "attributes": Array [],
  "children": Array [
    Object {
      "attributes": Array [],
      "children": Array [
        Object {
          "attributes": Array [],
          "children": Array [],
          "nodeName": "B",
          "onclick": undefined,
          "textContent": "foo",
        },
        Object {
          "attributes": Array [],
          "children": Array [],
          "nodeName": "B",
          "onclick": undefined,
          "textContent": "bar",
        },
        Object {
          "attributes": Array [],
          "children": Array [],
          "nodeName": "B",
          "onclick": undefined,
          "textContent": "baz",
        },
        Object {
          "attributes": Array [],
          "children": Array [],
          "nodeName": "B",
          "onclick": undefined,
          "textContent": "qux",
        },
      ],
      "nodeName": "P",
      "onclick": undefined,
      "textContent": "foobarbazqux",
    },
  ],
  "nodeName": "DIV",
  "onclick": undefined,
  "textContent": "foobarbazqux",
}
`;

exports[`clear a sequence 1`] = `
Object {
  "attributes": Array [],
  "children": Array [
    Object {
      "attributes": Array [],
      "children": Array [],
      "nodeName": "P",
      "onclick": undefined,
      "textContent": "",
    },
  ],
  "nodeName": "DIV",
  "onclick": undefined,
  "textContent": "",
}
`;

exports[`morph a component 1`] = `
Object {
  "attributes": Array [],
  "children": Array [
    Object {
      "attributes": Array [],
      "children": Array [],
      "nodeName": "P",
      "onclick": undefined,
      "textContent": "foo",
    },
  ],
  "nodeName": "DIV",
  "onclick": undefined,
  "textContent": "foo",
}
`;

exports[`morph a component into a primitive 1`] = `
Object {
  "attributes": Array [],
  "children": Array [
    Object {
      "attributes": Array [],
      "children": Array [],
      "nodeName": "P",
      "onclick": undefined,
      "textContent": "foo",
    },
  ],
  "nodeName": "DIV",
  "onclick": undefined,
  "textContent": "foo",
}
`;

exports[`morph a component into a sequence 1`] = `
Object {
  "attributes": Array [],
  "children": Array [
    Object {
      "attributes": Array [],
      "children": Array [],
      "nodeName": "P",
      "onclick": undefined,
      "textContent": "foobar",
    },
  ],
  "nodeName": "DIV",
  "onclick": undefined,
  "textContent": "foobar",
}
`;

exports[`morph a root component 1`] = `
Object {
  "attributes": Array [],
  "children": Array [
    Object {
      "attributes": Array [],
      "children": Array [],
      "nodeName": "I",
      "onclick": undefined,
      "textContent": "foo",
    },
  ],
  "nodeName": "DIV",
  "onclick": undefined,
  "textContent": "foo",
}
`;

exports[`morph a sequence 1`] = `
Object {
  "attributes": Array [],
  "children": Array [
    Object {
      "attributes": Array [],
      "children": Array [
        Object {
          "attributes": Array [],
          "children": Array [],
          "nodeName": "I",
          "onclick": undefined,
          "textContent": "foo",
        },
        Object {
          "attributes": Array [],
          "children": Array [],
          "nodeName": "I",
          "onclick": undefined,
          "textContent": "bar",
        },
      ],
      "nodeName": "P",
      "onclick": undefined,
      "textContent": "foobar",
    },
  ],
  "nodeName": "DIV",
  "onclick": undefined,
  "textContent": "foobar",
}
`;

exports[`morph a sequence into a component 1`] = `
Object {
  "attributes": Array [],
  "children": Array [
    Object {
      "attributes": Array [],
      "children": Array [
        Object {
          "attributes": Array [],
          "children": Array [],
          "nodeName": "B",
          "onclick": undefined,
          "textContent": "foo",
        },
      ],
      "nodeName": "P",
      "onclick": undefined,
      "textContent": "foo",
    },
  ],
  "nodeName": "DIV",
  "onclick": undefined,
  "textContent": "foo",
}
`;

exports[`morph a sequence into a primitive 1`] = `
Object {
  "attributes": Array [],
  "children": Array [
    Object {
      "attributes": Array [],
      "children": Array [],
      "nodeName": "P",
      "onclick": undefined,
      "textContent": "foo",
    },
  ],
  "nodeName": "DIV",
  "onclick": undefined,
  "textContent": "foo",
}
`;

exports[`remove items from a sequence 1`] = `
Object {
  "attributes": Array [],
  "children": Array [
    Object {
      "attributes": Array [],
      "children": Array [
        Object {
          "attributes": Array [],
          "children": Array [],
          "nodeName": "B",
          "onclick": undefined,
          "textContent": "foo",
        },
        Object {
          "attributes": Array [],
          "children": Array [],
          "nodeName": "B",
          "onclick": undefined,
          "textContent": "baz",
        },
      ],
      "nodeName": "P",
      "onclick": undefined,
      "textContent": "foobaz",
    },
  ],
  "nodeName": "DIV",
  "onclick": undefined,
  "textContent": "foobaz",
}
`;

exports[`reorder items of a sequence 1`] = `
Object {
  "attributes": Array [],
  "children": Array [
    Object {
      "attributes": Array [],
      "children": Array [
        Object {
          "attributes": Array [],
          "children": Array [],
          "nodeName": "B",
          "onclick": undefined,
          "textContent": "qux",
        },
        Object {
          "attributes": Array [],
          "children": Array [],
          "nodeName": "B",
          "onclick": undefined,
          "textContent": "baz",
        },
        Object {
          "attributes": Array [],
          "children": Array [],
          "nodeName": "B",
          "onclick": undefined,
          "textContent": "bar",
        },
        Object {
          "attributes": Array [],
          "children": Array [],
          "nodeName": "B",
          "onclick": undefined,
          "textContent": "foo",
        },
      ],
      "nodeName": "P",
      "onclick": undefined,
      "textContent": "quxbazbarfoo",
    },
  ],
  "nodeName": "DIV",
  "onclick": undefined,
  "textContent": "quxbazbarfoo",
}
`;

exports[`update items of a sequence 1`] = `
Object {
  "attributes": Array [],
  "children": Array [
    Object {
      "attributes": Array [],
      "children": Array [
        Object {
          "attributes": Array [],
          "children": Array [],
          "nodeName": "B",
          "onclick": undefined,
          "textContent": "baz",
        },
        Object {
          "attributes": Array [],
          "children": Array [],
          "nodeName": "B",
          "onclick": undefined,
          "textContent": "qux",
        },
      ],
      "nodeName": "P",
      "onclick": undefined,
      "textContent": "bazqux",
    },
  ],
  "nodeName": "DIV",
  "onclick": undefined,
  "textContent": "bazqux",
}
`;
